File: assignment-3/thomas/Assignment3.py
import traceback

def secondFunc():
    raise Exception("Test error in secondFunc")

def firstFunc():
    try:
        secondFunc()
    except Exception:
        raise Exception("Test error in firstFunc")
    
def modFirstFunc():
    try:
        secondFunc()
    except Exception as e:
        raise Exception("Test error in modFirstFunc") from e

def main():
    print('--- Starting firstFunc error handling ---')
    try:
        firstFunc()
    except Exception as e:
        print(f"Error: {e}")
        print(f"Cause: {e.__cause__}")
        print(f"Context: {e.__context__}")
        print("Stack Trace:")
        traceback.print_exc()
    print('--- End of firstFunc error handling ---\n')

    print('--- Starting modFirstFunc error handling ---')
    try:
        modFirstFunc()
    except Exception as e:
        print(f"Error: {e}")
        print(f"Cause: {e.__cause__}")
        print(f"Context: {e.__context__}")
        print("Stack Trace:")
        traceback.print_exc()
    print('--- End of modFirstFunc error handling ---')

if __name__ == "__main__":
    main()

File: assignment-3/thomas/java/src/main/java/com/thomas/Assignment3.java
package com.thomas;

public class Assignment3 {

    public static void secondFunc() throws Exception {
        throw new Exception("Test error in secondFunc");
    }

    public static void firstFunc() throws Exception {
        try {
            secondFunc();
        } catch (Exception e) {
            throw new Exception("Test error in firstFunc");
        }
    }

    public static void modFirstFunc() throws Exception {
        try {
            secondFunc();
        } catch (Exception e) {
            throw new Exception("Test error in modFirstFunc", e);
        }
    }

    public static void main(String[] args) {
        System.out.println("--- Starting firstFunc error handling ---");
        try {
            firstFunc();
        } catch (Exception e) {
            System.out.println("Error: " + e);
            System.out.println("Cause: " + e.getCause());
            System.out.println("Stack Trace:");
            e.printStackTrace(System.out);
        }
        System.out.println("--- End of firstFunc error handling ---\n");

        System.out.println("--- Starting modFirstFunc error handling ---");
        try {
            modFirstFunc();
        } catch (Exception e) {
            System.out.println("Error: " + e);
            System.out.println("Cause: " + e.getCause());
            System.out.println("Stack Trace:");
            e.printStackTrace(System.out);
        }
        System.out.println("--- End of modFirstFunc error handling ---");
    }
}
